### ISSUE TITLE
Investigate fsmonitor problem

### ISSUE THREAD

**Issue by dscho** [git-for-windows/git#2129]
The PR build of https://github.com/git-for-windows/git/pull/2127 failed a couple of times, always in the same test case where it tries to make sure that the fsmonitor hook prevents unnecessary `lstat()` calls. Re-running "fixed" this, but we still should investigate further. From [the log](https://dev.azure.com/Git-for-Windows/git/_build/results?buildId=33141):

```
Initialized empty Git repository in D:/a/1/s/t/trash directory.t7519-status-fsmonitor/.git/
expecting success: 
	mkdir -p .git/hooks &&
	: >tracked &&
	: >modified &&
	mkdir dir1 &&
	: >dir1/tracked &&
	: >dir1/modified &&
	mkdir dir2 &&
	: >dir2/tracked &&
	: >dir2/modified &&
	git -c core.fsmonitor= add . &&
	git -c core.fsmonitor= commit -m initial &&
	git config core.fsmonitor .git/hooks/fsmonitor-test &&
	cat >.gitignore <<-\EOF
	.gitignore
	expect*
	actual*
	marker*
	EOF

++ mkdir -p .git/hooks
++ :
++ :
++ mkdir dir1
++ :
++ :
++ mkdir dir2
++ :
++ :
++ git -c core.fsmonitor= add .
++ git -c core.fsmonitor= commit -m initial
[master (root-commit) 7fbd2dc] initial
 Author: A U Thor <author@example.com>
 6 files changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 dir1/modified
 create mode 100644 dir1/tracked
 create mode 100644 dir2/modified
 create mode 100644 dir2/tracked
 create mode 100644 modified
 create mode 100644 tracked
++ git config core.fsmonitor .git/hooks/fsmonitor-test
++ cat
ok 1 - setup

expecting success: 
	test-tool dump-fsmonitor >actual &&
	grep "^no fsmonitor" actual

++ test-tool dump-fsmonitor
++ grep '^no fsmonitor' actual
no fsmonitor

ok 2 - fsmonitor extension is off by default
expecting success: 
	git update-index --fsmonitor &&
	test-tool dump-fsmonitor >actual &&
	grep "^fsmonitor last update" actual

++ git update-index --fsmonitor
error: cannot spawn .git/hooks/fsmonitor-test: No such file or directory
++ test-tool dump-fsmonitor
++ grep '^fsmonitor last update' actual
fsmonitor last update 1552652689120621600
ok 3 - update-index --fsmonitor" adds the fsmonitor extension

expecting success: 
	git update-index --no-fsmonitor &&
	test-tool dump-fsmonitor >actual &&
	grep "^no fsmonitor" actual

++ git update-index --no-fsmonitor
warning: core.fsmonitor is set; remove it if you really want to disable fsmonitor
++ test-tool dump-fsmonitor
++ grep '^no fsmonitor' actual
no fsmonitor
ok 4 - update-index --no-fsmonitor" removes the fsmonitor extension

expecting success: 
	git update-index --fsmonitor &&
	git update-index --fsmonitor-valid dir1/modified &&
	git update-index --fsmonitor-valid dir2/modified &&
	git update-index --fsmonitor-valid modified &&
	git ls-files -f >actual &&
	test_cmp expect actual

++ git update-index --fsmonitor
error: cannot spawn .git/hooks/fsmonitor-test: No such file or directory
++ git update-index --fsmonitor-valid dir1/modified
++ git update-index --fsmonitor-valid dir2/modified
++ git update-index --fsmonitor-valid modified
++ git ls-files -f
++ test_cmp expect actual
++ GIT_ALLOC_LIMIT=0
++ test-tool cmp expect actual

ok 5 - update-index --fsmonitor-valid" sets the fsmonitor valid bit
expecting success: 
	git update-index --no-fsmonitor-valid dir1/modified &&
	git update-index --no-fsmonitor-valid dir2/modified &&
	git update-index --no-fsmonitor-valid modified &&
	git ls-files -f >actual &&
	test_cmp expect actual

++ git update-index --no-fsmonitor-valid dir1/modified
++ git update-index --no-fsmonitor-valid dir2/modified
++ git update-index --no-fsmonitor-valid modified
++ git ls-files -f
++ test_cmp expect actual
++ GIT_ALLOC_LIMIT=0
++ test-tool cmp expect actual
ok 6 - update-index --no-fsmonitor-valid" clears the fsmonitor valid bit

expecting success: 
	write_integration_script &&
	dirty_repo &&
	git update-index --fsmonitor &&
	git ls-files -f >actual &&
	test_cmp expect actual

++ write_integration_script
++ write_script .git/hooks/fsmonitor-test
++ echo '#!/bin/sh'
++ cat
++ chmod +x .git/hooks/fsmonitor-test
++ dirty_repo
++ :
++ :
++ :
++ echo 1
++ echo 2
++ echo 3
++ echo 4
++ echo 5
++ echo 6
++ git update-index --fsmonitor
++ git ls-files -f
++ test_cmp expect actual
++ GIT_ALLOC_LIMIT=0
++ test-tool cmp expect actual

ok 7 - all files returned by integration script get flagged as invalid
expecting success: 
	git add new &&
	git add dir1/new &&
	git add dir2/new &&
	git ls-files -f >actual &&
	test_cmp expect actual

++ git add new
++ git add dir1/new
++ git add dir2/new
++ git ls-files -f
++ test_cmp expect actual
++ GIT_ALLOC_LIMIT=0
++ test-tool cmp expect actual

ok 8 - newly added files are marked valid
expecting success: 
	# modified files result in update-index returning 1
	test_must_fail git update-index --refresh --force-write-index &&
	git ls-files -f >actual &&
	test_cmp expect actual

++ test_must_fail git update-index --refresh --force-write-index
++ case "$1" in
++ _test_ok=
++ git update-index --refresh --force-write-index
dir1/modified: needs update
dir2/modified: needs update
modified: needs update
++ exit_code=1
++ test 1 -eq 0
++ test_match_signal 13 1
++ test 1 = 141
++ test 1 = 269
++ return 1
++ test 1 -gt 129
++ test 1 -eq 127
++ test 1 -eq 126
++ return 0
++ git ls-files -f
++ test_cmp expect actual
++ GIT_ALLOC_LIMIT=0
++ test-tool cmp expect actual
ok 9 - all unmodified files get marked valid

expecting success: 
	write_script .git/hooks/fsmonitor-test<<-\EOF &&
	printf "dir1/modified\0"
	EOF
	clean_repo &&
	git update-index --refresh --force-write-index &&
	echo 1 >modified &&
	echo 2 >dir1/modified &&
	echo 3 >dir2/modified &&
	test_must_fail git update-index --refresh --force-write-index &&
	git ls-files -f >actual &&
	test_cmp expect actual

++ write_script .git/hooks/fsmonitor-test
++ echo '#!/bin/sh'
++ cat
++ chmod +x .git/hooks/fsmonitor-test
++ clean_repo
++ git reset --hard HEAD
HEAD is now at 7fbd2dc initial
++ git clean -fd
Removing dir1/untracked
Removing dir2/untracked
Removing untracked
++ git update-index --refresh --force-write-index
++ echo 1
++ echo 2
++ echo 3
++ test_must_fail git update-index --refresh --force-write-index
++ case "$1" in
++ _test_ok=
++ git update-index --refresh --force-write-index
dir1/modified: needs update
++ exit_code=1
++ test 1 -eq 0
++ test_match_signal 13 1
++ test 1 = 141
++ test 1 = 269
++ return 1
++ test 1 -gt 129
++ test 1 -eq 127
++ test 1 -eq 126
++ return 0
++ git ls-files -f
++ test_cmp expect actual
++ GIT_ALLOC_LIMIT=0
++ test-tool cmp expect actual
ok 10 - *only* files returned by the integration script get flagged as invalid

expecting success: 
	write_script .git/hooks/fsmonitor-test<<-\EOF &&
	EOF
	clean_repo &&
	dirty_repo &&
	git add . &&
	git commit -m "to reset" &&
	git reset HEAD~1 &&
	git status >actual &&
	git -c core.fsmonitor= status >expect &&
	test_i18ncmp expect actual

++ write_script .git/hooks/fsmonitor-test
++ echo '#!/bin/sh'
++ cat
++ chmod +x .git/hooks/fsmonitor-test
++ clean_repo
++ git reset --hard HEAD
HEAD is now at 7fbd2dc initial
++ git clean -fd
++ dirty_repo
++ :
++ :
++ :
++ echo 1
++ echo 2
++ echo 3
++ echo 4
++ echo 5
++ echo 6
++ git add .
++ git commit -m 'to reset'
[master 27bdddb] to reset
 Author: A U Thor <author@example.com>
 9 files changed, 6 insertions(+)
 create mode 100644 dir1/new
 create mode 100644 dir1/untracked
 create mode 100644 dir2/new
 create mode 100644 dir2/untracked
 create mode 100644 new
 create mode 100644 untracked
++ git reset HEAD~1
Unstaged changes after reset:
M	dir1/modified
M	dir2/modified
M	modified
++ git status
++ git -c core.fsmonitor= status
++ test_i18ncmp expect actual
++ test_have_prereq C_LOCALE_OUTPUT
++ save_IFS=' 	
'
++ IFS=,
++ set -- C_LOCALE_OUTPUT
++ IFS=' 	
'
++ total_prereq=0
++ ok_prereq=0
++ missing_prereq=
++ for prerequisite in "$@"
++ case "$prerequisite" in
++ negative_prereq=
++ case " $lazily_tested_prereq " in
++ case " $lazily_testable_prereq " in
++ total_prereq=1
++ case "$satisfied_prereq" in
++ satisfied_this_prereq=t
++ case "$satisfied_this_prereq,$negative_prereq" in
++ ok_prereq=1
++ test 1 = 1
++ test_cmp expect actual
++ GIT_ALLOC_LIMIT=0
++ test-tool cmp expect actual
ok 11 - refresh_index() invalidates fsmonitor cache

expecting success: 
		git config core.preloadIndex $preload_val &&
		if test $preload_val = true
		then
			GIT_TEST_PRELOAD_INDEX=$preload_val; export GIT_TEST_PRELOAD_INDEX
		else
			sane_unset GIT_TEST_PRELOAD_INDEX
		fi
	
++ git config core.preloadIndex false
++ test false = true
++ sane_unset GIT_TEST_PRELOAD_INDEX
++ unset GIT_TEST_PRELOAD_INDEX
++ return 0
ok 12 - setup preloadIndex to false

checking prerequisite: UNTRACKED_CACHE

mkdir -p "$TRASH_DIRECTORY/prereq-test-dir" &&
(
	cd "$TRASH_DIRECTORY/prereq-test-dir" &&
	{ git update-index --test-untracked-cache; ret=$?; } &&
	test $ret -ne 1

)
++ mkdir -p '/d/a/1/s/t/trash directory.t7519-status-fsmonitor/prereq-test-dir'
++ cd '/d/a/1/s/t/trash directory.t7519-status-fsmonitor/prereq-test-dir'
++ git update-index --test-untracked-cache
Testing mtime in 'D:/a/1/s/t/trash directory.t7519-status-fsmonitor' ...... OK
++ ret=0
++ test 0 -ne 1
prerequisite UNTRACKED_CACHE ok
expecting success: 
			git config core.untrackedcache $uc_val
		
++ git config core.untrackedcache false

ok 13 - setup untracked cache to false
expecting success: 
			write_integration_script &&
			clean_repo &&
			dirty_repo &&
			git add new &&
			git add dir1/new &&
			git add dir2/new &&
			git status >actual &&
			git -c core.fsmonitor= status >expect &&
			test_i18ncmp expect actual
		
++ write_integration_script
++ write_script .git/hooks/fsmonitor-test
++ echo '#!/bin/sh'
++ cat
++ chmod +x .git/hooks/fsmonitor-test
++ clean_repo
++ git reset --hard HEAD
HEAD is now at 7fbd2dc initial
++ git clean -fd
Removing dir1/new
Removing dir1/untracked
Removing dir2/new
Removing dir2/untracked
Removing new
Removing untracked
++ dirty_repo
++ :
++ :
++ :
++ echo 1
++ echo 2
++ echo 3
++ echo 4
++ echo 5
++ echo 6
++ git add new
++ git add dir1/new
++ git add dir2/new
++ git status
++ git -c core.fsmonitor= status
++ test_i18ncmp expect actual
++ test_have_prereq C_LOCALE_OUTPUT
++ save_IFS=' 	
'
++ IFS=,
++ set -- C_LOCALE_OUTPUT
++ IFS=' 	
'
++ total_prereq=0
++ ok_prereq=0
++ missing_prereq=
++ for prerequisite in "$@"
++ case "$prerequisite" in
++ negative_prereq=
++ case " $lazily_tested_prereq " in
++ case " $lazily_testable_prereq " in
++ total_prereq=1
++ case "$satisfied_prereq" in
++ satisfied_this_prereq=t
++ case "$satisfied_this_prereq,$negative_prereq" in
++ ok_prereq=1
++ test 1 = 1
++ test_cmp expect actual
++ GIT_ALLOC_LIMIT=0
++ test-tool cmp expect actual

ok 14 - compare status with and without fsmonitor
expecting success: 
			write_script .git/hooks/fsmonitor-test<<-\EOF &&
			:>marker
			EOF
			clean_repo &&
			git status &&
			test_path_is_file marker &&
			dirty_repo &&
			rm -f marker &&
			git status >actual &&
			test_path_is_file marker &&
			test_i18ngrep ! "Changes not staged for commit:" actual &&
			if test $uc_val = true
			then
				test_i18ngrep ! "Untracked files:" actual
			fi &&
			if test $uc_val = false
			then
				test_i18ngrep "Untracked files:" actual
			fi &&
			rm -f marker
		
++ write_script .git/hooks/fsmonitor-test
++ echo '#!/bin/sh'
++ cat
++ chmod +x .git/hooks/fsmonitor-test
++ clean_repo
++ git reset --hard HEAD
HEAD is now at 7fbd2dc initial
++ git clean -fd
Removing dir1/untracked
Removing dir2/untracked
Removing untracked
++ git status
On branch master
nothing to commit, working tree clean
++ test_path_is_file marker
++ test -f marker
++ dirty_repo
++ :
++ :
++ :
++ echo 1
++ echo 2
++ echo 3
++ echo 4
++ echo 5
++ echo 6
++ rm -f marker
++ git status
++ test_path_is_file marker
++ test -f marker
++ test_i18ngrep '!' 'Changes not staged for commit:' actual
++ eval 'last_arg=${3}'
+++ last_arg=actual
++ test -f actual
++ test 3 -lt 2
++ test 'x!' = 'x!'
++ test 3 -lt 3
++ test_have_prereq '!C_LOCALE_OUTPUT'
++ save_IFS=' 	
'
++ IFS=,
++ set -- '!C_LOCALE_OUTPUT'
++ IFS=' 	
'
++ total_prereq=0
++ ok_prereq=0
++ missing_prereq=
++ for prerequisite in "$@"
++ case "$prerequisite" in
++ negative_prereq=t
++ prerequisite=C_LOCALE_OUTPUT
++ case " $lazily_tested_prereq " in
++ case " $lazily_testable_prereq " in
++ total_prereq=1
++ case "$satisfied_prereq" in
++ satisfied_this_prereq=t
++ case "$satisfied_this_prereq,$negative_prereq" in
++ prerequisite='!C_LOCALE_OUTPUT'
++ test -z ''
++ missing_prereq='!C_LOCALE_OUTPUT'
++ test 1 = 0
++ test 'x!' = 'x!'
++ shift
++ grep 'Changes not staged for commit:' actual
Changes not staged for commit:
++ echo 'error: '\''! grep Changes not staged for commit:' 'actual'\'' did find a match in:'
error: '! grep Changes not staged for commit: actual' did find a match in:
++ test -s actual
++ cat actual
On branch master
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   dir1/modified
	modified:   dir2/modified
	modified:   modified

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	dir1/new
	dir1/untracked
	dir2/new
	dir2/untracked
	new
	untracked

no changes added to commit (use "git add" and/or "git commit -a")
++ return 1
error: last command exited with $?=1
```

**Comment by dscho** [comment]
I should add that I tried to reproduce this locally, failing many times. I even ran `--stress --stress-limit=50`, but I did get the impression that this is more of a "light load" problem, i.e. some sort of a racy timestamp "so I ignore if the timestamp suggests that things are up to date" issue.

**Comment by dscho** [comment]
I _think_ that I fixed this in 398a3b0899dd8a440d4adbcbda38362e3f8359b1. In any case, the test case no longer fails.
